feat(api): DNS challenge support for Let's Encrypt

nils-wisiol opened the pull request:
Provides easy setting of TXT records at `_acme-challenge.yourname.dedyn.io.`

A certbot hook script can be found at https://github.com/desec-io/certbot-hook

peterthomassen commented on api/desecapi/migrations/0013_acme_challenge.py:
```diff
@@ -0,0 +1,20 @@
+        migrations.AddField(
+            model_name='domain',
+            name='acme_challenge',
+            field=models.CharField(max_length=1024, blank=True),
```
The length of a string in a `TXT` RR is limited to 255, so this field is too wide. I leave if up to your judgment to reflect that fact here or not.

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -99,6 +99,7 @@ class Domain(models.Model):
     arecord = models.CharField(max_length=255, blank=True)
     aaaarecord = models.CharField(max_length=1024, blank=True)
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
+    acme_challenge = models.CharField(max_length=1024, blank=True)
```
See above

peterthomassen commented on api/desecapi/tests/testdomains.py:
```diff
@@ -189,13 +199,29 @@ def testPostingWithRecordsCausesPdnsAPIPatch(self):
         self.assertTrue('dead::beef' in httpretty.last_request().parsed_body)
+        self.assertTrue('letsencrypt_ftw' in httpretty.last_request().parsed_body)
+
+    def testPatchDomainCausesPdnsAPIPatch(self):
```
This is a misnomer (you are posting to the domain, not patching)

nils-wisiol commented on api/desecapi/migrations/0013_acme_challenge.py:
```diff
@@ -0,0 +1,20 @@
+        migrations.AddField(
+            model_name='domain',
+            name='acme_challenge',
+            field=models.CharField(max_length=255, blank=True),
```
length set to 255, however it is unclear if bytes or characters are counted (UTF-8)

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -99,6 +99,7 @@ class Domain(models.Model):
     arecord = models.CharField(max_length=255, blank=True)
     aaaarecord = models.CharField(max_length=1024, blank=True)
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
+    acme_challenge = models.CharField(max_length=255, blank=True)
```
updating any db that was build with the previous version of this PR will fail, drop database then rebuild from scratch (or change manually).

nils-wisiol commented on api/desecapi/tests/testdomains.py:
```diff
@@ -189,13 +199,29 @@ def testPostingWithRecordsCausesPdnsAPIPatch(self):
         self.assertTrue('dead::beef' in httpretty.last_request().parsed_body)
+        self.assertTrue('letsencrypt_ftw' in httpretty.last_request().parsed_body)
+
+    def testPostDomainCausesPdnsAPIPatch(self):
```
notice `post`

nils-wisiol commented:
I addressed all requested changes; in an effort to streamline the review process I also added comments on the sections in the code that I changed.

I also first pushed with fixup commits, then rebased and squashed them and push --force again. (When trying how GitHub would deal with the outdated diffs in that case -- the handling did not improve.) That's how there is several commits showing on this page; but only one commit showing in the PR itself.
